// @uppy/companion/lib/server/provider/Provider.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/**
 * Provider interface defines the specifications of any provider implementation
 */
class Provider {
    /**
     *
     * @param {object} options
     */
    constructor(options) {
        this.needsCookieAuth = false;
        return this;
    }
    /**
     * config to extend the grant config
     */
    static getExtraConfig() {
        return {};
    }
    /**
     * list the files and folders in the provider account
     *
     * @param {object} options
     * @returns {Promise}
     */
    // eslint-disable-next-line class-methods-use-this,no-unused-vars
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error('method not implemented');
        });
    }
    /**
     * download a certain file from the provider account
     *
     * @param {object} options
     * @returns {Promise}
     */
    // eslint-disable-next-line class-methods-use-this,no-unused-vars
    download(options) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error('method not implemented');
        });
    }
    /**
     * return a thumbnail for a provider file
     *
     * @param {object} options
     * @returns {Promise}
     */
    // eslint-disable-next-line class-methods-use-this,no-unused-vars
    thumbnail(options) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error('method not implemented');
        });
    }
    /**
     * get the size of a certain file in the provider account
     *
     * @param {object} options
     * @returns {Promise}
     */
    // eslint-disable-next-line class-methods-use-this,no-unused-vars
    size(options) {
        return __awaiter(this, void 0, void 0, function* () {
            throw new Error('method not implemented');
        });
    }
    /**
     * handle deauthorization notification from oauth providers
     *
     * @param {object} options
     * @returns {Promise}
     */
    // eslint-disable-next-line class-methods-use-this,no-unused-vars
    deauthorizationCallback(options) {
        return __awaiter(this, void 0, void 0, function* () {
            // @todo consider doing something like throw new NotImplementedError() instead
            throw new Error('method not implemented');
        });
    }
    /**
     * Name of the OAuth provider. Return empty string if no OAuth provider is needed.
     *
     * @returns {string}
     */
    static get authProvider() {
        return undefined;
    }
}
module.exports = Provider;
// OAuth providers are those that have a `static authProvider` set. It means they require OAuth authentication to work
module.exports.isOAuthProvider = (authProvider) => typeof authProvider === 'string' && authProvider.length > 0;
